Initial WG LTS Reformation (#7287)
kubernetes/community: wg-lts/charter.md

* constraints on deployment and upgrade patterns / timelines (e.g. edge deployments, regulated industries, retail, etc)
* expected/required support periods from users/vendors
* support periods of core Kubernetes dependencies
* support periods of other components required to run Kubernetes clusters (OS, network, storage, etc)
* what users/vendors are currently doing to support releases past OSS End Of Life (EOL)
* what do users/vendors do with EOL clusters today? what would they do if EOL was extended by N months/years?

With this information, we will investigate changes the Kubernetes project could make to address these needs. A non-exhaustive set of changes to investigate include:
* lengthening support period for all minor versions
* lengthening support period for specific minor versions
* applying security fixes to release branches past current EOL without cutting additional patch releases
* applying security and "critical" fixes to release branches past current EOL without cutting additional patch releases
* expanding supported skew
* improving supported upgrade patterns for clusters at EOL

> ### In Scope

Reviewer: Can you please consider some more of the items from the earlier WG charter? https://github.com/kubernetes/community/blob/master/archive/wg-lts/charter.md#in-scope

Reviewer 2: Are there any specific ones still missing at this point you'd like included?

- Collecting input to better define long term support with regard to Kubernetes OSS releases
- Creating and prioritizing a list of areas that require investments to improve long term supportability
- Initiate and drive cross-SIG changes related to long term support

> ### Out of scope

Reviewer: please consider adding at least the first bullet of the previous charter's out-of-scope section. https://github.com/kubernetes/community/blob/master/archive/wg-lts/charter.md#out-of-scope

Reviewer 3: Resolving this since it's added.

## Special Powers

None

## Stakeholders

Any changes identified by this working group will require involvement and investment from the following SIGS:

- SIG Architecture
  High-level input on requirements.
- SIG Cluster Lifecycle
  Input on cluster upgrade mechanics.
- SIG K8s Infra
  Input on infrastructure costs related to increased support period(s).
- SIG Release
  Input on maintaining older branches and additional releases.
- SIG Testing
  Input on testing impacts related to upgrades and longer test periods. 
  
Improvements identified will likely involve every SIG, but the list above are identified as the primary stakeholders.

## Deliverables

The artifacts the group is supposed to deliver include:
- Survey results better describing Kubernetes version use patterns, deployment constraints, and upgrade patterns.
- Recommend changes Kubernetes SIGS can make that will provide broad benefits in a sustainable/affordable way. This would likely take the form of one or more KEPs.
- Recommend ways users/vendors who want to maximize Kubernetes support can consume Kubernetes

Any changes identified will be owned by corresponding SIGs.

## Roles and Organization Management

> This sig follows adheres to the Roles and Organization Management outlined in

Reviewer 4: ```suggestion
This working group adheres to the Roles and Organization Management outlined in
```

[sig-governance] and opts-in to updates and modifications to [sig-governance].

[sig-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-governance.md

## Timelines and Disbanding

> The exact timeline for this working group is hard to define at this time. If we are unable to define a common long term support definition for Kubernetes, we will disband the working group. If we are unable to define improvements related to an agreed upon definition, we will also disband the working group.

Reviewer 5: Should we write periodic updates and reviews to sig-arch/steering/someone else in here? Perhaps a 6-monthly update, and a review after 12 months?

Feels like having some deadlines to produce something might help keep the discussion focussed.

Author: @youngnick  that's a good suggestion, I will incorporate your feedback.

# WG Long Term Support

This charter adheres to the conventions described in the [Kubernetes Charter README]
and uses the Roles and Organization Management outlined in [sig-governance].

[sig-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-governance.md
[Kubernetes Charter README]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/README.md

## Scope

The Long Term Support Working Group (WG LTS) is organized with the goal of developing a better understanding what "Long Term Support" might mean for Kubernetes, those who support Kubernetes, and end users, as well as to investigate changes the Kubernetes project could make related to long term support. The working group will also determine feasibility, benefits, **costs**, and prerequisites of such changes.

In the first phase of the working group, we will collect information related to the needs and wants regarding support periods from end-users and people supporting Kubernetes. This will be done by reaching out to users and cluster operators (e.g. via surveys), to better understand:

> * in-use versions (in the form of a user/vendor survey) and reasons for remaining on those versions

Reviewer 6: For the survey, we have done a survey of our customers that is similar.
- The survey should cover that users may have multi clusters in multi versions. The cluster size and number of clusters also matter for upgradation. For instance, migration is another approach for multi-clusters. Some big clusters may be hard to upgrade due to complex configurations and some special dependencies(storage/networking) of pods. Some customers chose to drain old-version clusters like draining a node. `Drain old version cluster` means to migrate apps to new clusters when apps need an update.
-  How often does your cluster upgrade minor versions? How often does your cluster upgrade patch versions?

Reviewer 7: It's a little early to describe all survey questions in this initial reformation PR. Some examples are good enough.

Admit that survey questions will impact the WG-LTS conclusion greatly, need to be designed and reviewed carefully.

for exmple, if the question is expected support period for each minor version, may receive answer "the longer the better". But if the question is can you accept only one longer supported minor version every year, the answer maybe "yes".

Existing facts will be easier to answer and expressed exactly. Expectation is not equal to real need, sometimes will be overstated.

Reviewer 8: Agreed - probably don't need to expand too much on survey questions here. It's fine to list a few as examples, but there will likely be a significant refinement of them. We do have some data already from the regular Production Readiness Surveys.

* constraints on deployment and upgrade patterns / timelines (e.g. edge deployments, regulated industries, retail, etc)
* expected/required support periods from users/vendors
* support periods of core Kubernetes dependencies
* support periods of other components required to run Kubernetes clusters (OS, network, storage, etc)
* what users/vendors are currently doing to support releases past community support End Of Life (EOL)

> * what do users/vendors do with EOL clusters today? what would they do if EOL was extended by N months/years?

Reviewer 6: > EOL was extended by N months/years.

Extending the current long-term support of all versions?

As I know, some other software will mark some versions as long support and others as not.
- for instance, we may mark one release of a year as long support for some extended time only. This may save some effort of the community.

Reviewer 9: That isn't known yet, that's one of the questions to investigate/answer

* lengthening support period for all minor versions
* lengthening support period for specific minor versions
* applying security fixes to release branches past current EOL without cutting additional patch releases
* applying security and "critical" fixes to release branches past current EOL without cutting additional patch releases
* expanding supported skew
* improving supported upgrade patterns for clusters at EOL

### In Scope

- Collecting input to better define long term support with regard to Kubernetes releases. This could include:
  * What is a supported release?
  * Number of community supported branches.
  * Duration of community support per supported branch.
- Working with vendors and other community members to identify ongoing efforts to help end users beyond current community support and identify opportunities for reuse of prior work.
- Creating and prioritizing a list of areas that require investments to improve long term supportability. This could include:
  * Upgrade path considerations.
  * Costs of Kubernetes releases in terms of:
    * Infrastructure
    * People
- Initiate and drive cross-SIG changes related to long term support.

> - Identify ways to that vendors and other community members can better support Kubernetes releases.

Author: Thanks!

  responsibilities. This is a working group, no code implementation is the responsibility of this Working Group.
- Technical and end-user support:  The WG may make recommendations
  around support to those responsible for relevant code and responsible
  for the release engineering operations and automation, but does not
  own code itself.

## Special Powers

None

## Stakeholders

Any changes identified by this working group will require involvement and investment from the following SIGS:

- SIG Architecture
  High-level input on requirements.
- SIG Cluster Lifecycle
  Input on cluster upgrade mechanics.
- SIG K8s Infra
  Input on infrastructure costs related to increased support period(s).
- SIG Release
  Input on maintaining older branches and additional releases.
- SIG Testing
  Input on testing impacts related to upgrades and longer test periods. 
  
Improvements identified will likely involve every SIG, but the list above are identified as the primary stakeholders.

## Deliverables

The artifacts the group is supposed to deliver include:
- Survey results better describing Kubernetes version use patterns, deployment constraints, and upgrade patterns.

> - Recommend changes Kubernetes SIGS can make that will provide broad benefits in a sustainable/affordable way. This would likely take the form of one or more KEPs.

Reviewer 3: nit:
```suggestion
- Recommend changes Kubernetes SIGs can make that will provide broad benefits in a sustainable/affordable way. This would likely take the form of one or more KEPs.
```

> - Recommend ways users/vendors who want to maximize Kubernetes support can consume Kubernetes

Reviewer 8: Generally if the target audience is end users, docs are pushed to the primary website + blog article highlighting them.

For disruptive changes we do engage other communities more directly - e.g. pushing to reddit, CNCF end user group etc.

Any changes identified will be owned by corresponding SIGs.

## Roles and Organization Management

This working group adheres to the Roles and Organization Management outlined in
[sig-governance] and opts-in to updates and modifications to [sig-governance].

[sig-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-governance.md

## Timelines and Disbanding

> The exact timeline for this working group is hard to define at this time. If we are unable to define a common long term support definition for Kubernetes, we will disband the working group. If we are unable to define improvements related to an agreed upon definition, we will also disband the working group. In order to evaluate our progress toward the working group goals, we will provide periodic updates to the stakeholder SIGs and committees at least every six months, in addition to a working group annual report.   

Reviewer 10: Would the lack of people and higher costs for providing long term support be a reason for disbanding the working group?

Reviewer 8: I think that is more a potential outcome of the WG. If the WG determines we currently do not have the resources, nor commitments for future resources to potentially support an LTS then theres no reason to continue.

Author: I added this as reason for disbanding for clarity.

  * Costs of Kubernetes releases in terms of:
    * Infrastructure
    * People
- Initiating and driving cross-SIG changes related to long term support.
- Identifing ways to that vendors and other community members can better support Kubernetes releases.

### Out of scope

- The lifecycle of projects outside of the Kubernetes org.
- Designing and executing on changes clearly falling into individual SIG
  responsibilities. This is a working group, no code implementation is the responsibility of this Working Group.
- Technical and end-user support:  The WG may make recommendations
  around support to those responsible for relevant code and responsible
  for the release engineering operations and automation, but does not
  own code itself.

## Special Powers

None

## Stakeholders

Any changes identified by this working group will require involvement and investment from the following SIGS:

- SIG Architecture
  High-level input on requirements.
- SIG Cluster Lifecycle
  Input on cluster upgrade mechanics.
- SIG K8s Infra
  Input on infrastructure costs related to increased support period(s).
- SIG Release
  Input on maintaining older branches and additional releases.
- SIG Testing

>   Input on testing impacts related to upgrades and longer test periods. 

Reviewer 11: I see that SIG Security was added as a stakeholder, maybe an update here slipped through the cracks?

Author: Thanks @MadhavJivrajani, let me fix that up.

* improving supported upgrade patterns for clusters at EOL

### In Scope

- Collecting input to better define long term support with regard to Kubernetes releases. This could include:
  * What is a supported release?
  * Number of community supported branches.
  * Duration of community support per supported branch.
- Working with vendors and other community members to identify ongoing efforts to help end users beyond current community support and identify opportunities for reuse of prior work.
- Creating and prioritizing a list of areas that require investments to improve long term supportability. This could include:
  * Upgrade path considerations.
  * Costs of Kubernetes releases in terms of:
    * Infrastructure
    * People
- Initiating and driving cross-SIG changes related to long term support.
- Identifing ways to that vendors and other community members can better support Kubernetes releases.

### Out of scope

- The lifecycle of projects outside of the Kubernetes org.
- Designing and executing on changes clearly falling into individual SIG
  responsibilities. This is a working group, no code implementation is the responsibility of this Working Group.

> - Technical and end-user support:  The WG may make recommendations

Reviewer 12: ```suggestion
- Technical and end-user support: The WG may make recommendations
```
